### PR TITLE
[BUG] Fix failing tests due to tensorflow update

### LINKED ISSUE #6094: [BUG] Tests failing on main after recent `tensorflow` and `mlflow` releases
Ref. https://github.com/sktime/sktime/actions/runs/8211839419

FYI @achieveordie for mlflow error

> FAILED sktime/utils/tests/test_mlflow_sktime_model_export.py::test_cnn_model_save_and_load[pickle] - ValueError: Invalid filepath extension for saving. Please add either a `.keras` extension for the native Keras format (recommended) or a `.h5` extension. Use `tf.saved_model.save()` if you want to export a SavedModel for use with TFLite/TFServing/etc. Received: filepath=/tmp/pytest-of-runner/pytest-0/popen-gw0/test_cnn_model_save_and_load_p0/model/model/keras.
FAILED sktime/utils/tests/test_mlflow_sktime_model_export.py::test_cnn_model_save_and_load[cloudpickle] - ValueError: Invalid filepath extension for saving. Please add either a `.keras` extension for the native Keras format (recommended) or a `.h5` extension. Use `tf.saved_model.save()` if you want to export a SavedModel for use with TFLite/TFServing/etc. Received: filepath=/tmp/pytest-of-runner/pytest-0/popen-gw2/test_cnn_model_save_and_load_c0/model/model/keras.

FYI @fkiraly for classification error

> FAILED sktime/classification/tests/test_base.py::test_deep_estimator_full[object-adamax] - TypeError: 'list' object is not callable

### PR DESCRIPTION
**Pull request by benHeid**
#### Reference Issues/PRs
Fixes #6094 

#### What does this implement/fix? Explain your changes.
It uses the proposed save format of tensorflow. To be downward compatible, I introduced an if-else statement to select the save functionality. Should we warn the user and deprecate the old 

#### What should a reviewer concentrate their feedback on?
* Is the deprecation cycle appropriate? 
* With this change, the test will only be green for tf >= 2.16... Should I add a if-else case here or is it okay for the tests.

#### Did you add any tests for the change?

No

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
Hm, I see.

The issue is, the general `save` contract assumes that the function loads and writes to any file name, so forcing a file ending would violate that.

Would a solution that satisfies the contract not rather be this:

* save to a file with the ending
* rename the file

Or similar?

Further, the specification - and docstring - says the file should be a `zip`. I am a bit confused why the ending is required in the serialized file, rather than, say, in a file that is contained in the `zip`.

**Comment by benHeid** [comment]
> Hm, I see.
> 
> The issue is, the general `save` contract assumes that the function loads and writes to any file name, so forcing a file ending would violate that.
> 
> Would a solution that satisfies the contract not rather be this:
> 
> * save to a file with the ending
> * rename the file
> 
> Or similar?

I suppose this would be possible. However, I do not really like to do that, because I think it is very confusing, since file endings normally contains information on how to interpret a file and this would not be possible. 
 
> Further, the specification - and docstring - says the file should be a `zip`. I am a bit confused why the ending is required in the serialized file, rather than, say, in a file that is contained in the `zip`.

Currently, it can be a zip file if the version is lower than 2.16.0 and legacy_save=True. But I can change the docstrings..

**Comment by fkiraly** [comment]
> because I think it is very confusing, since file endings normally contains information on how to interpret a file and this would not be possible.

I think it would be more confusing to have different file endings depending on estimator - the whole idea is to abstract that away for the user, and it's always `zip`.

This has unfortunately not been written up properly, I cannot find it in the file specifications folder. @achieveordie, do you remember where the specification for the model serialization format went?

**Comment by fkiraly** [comment]
Found it - this should be moved to file specifications
https://github.com/sktime/enhancement-proposals/blob/main/steps/20_saveload/step.md

**Comment by fkiraly** [comment]
FYI @blazingbhavneek, @AurumnPegasus, in case you are still active.

**Comment by benHeid** [comment]
> I think it would be more confusing to have different file endings depending on estimator - the whole idea is to abstract that away for the user, and it's always `zip`.


Thus a solution then could be save it as .keras and putting this file into a zip?

**Comment by fkiraly** [comment]
> Thus a solution then could be save it as .keras and putting this file into a zip?

Yes - I'm slightly surprised that it wasn't already in a `zip` file, seems like an oversight related to not having written up the specs properly.

**Comment by benHeid** [comment]
Ok the remaining error seems to be caused by the in memory serialization. Unfortunately, I have no good idea how to solve it... Since `save` is only accepting str or Path objects and no files anymore as in the current code:

```python
if self.model_ is not None:
    with h5py.File(
        "disk_less", "w", driver="core", backing_store=False
    ) as h5file:
        self.model_.save(h5file)
        h5file.flush()
        in_memory_model = h5file.id.get_file_image()
```

File: sktime/classification/tests/test_base.py [hunk 1]
```diff
@@ -500,9 +500,7 @@ def test_deep_estimator_full(optimizer):
         assert serialize(full_dummy.optimizer) == serialize(deserialized_full.optimizer)
 
         # assert weights of optimizers are same
-        assert (
-            full_dummy.optimizer.variables() == deserialized_full.optimizer.variables()
-        )
+        assert full_dummy.optimizer.variables == deserialized_full.optimizer.variables
```
[the remark this answers is not in the document]
Comment (benHeid) [reply sktime/classification/tests/test_base.py:503]: Probably yes. However, this is part of the tests and the tests are always executed with the newest packages. Thus, I haven't added that check.

**Review by fkiraly — DISMISSED** [review]
Approved under the assumption that you tested it.

What I'm a bit surprised about is that there's no `legacy_save` method distinction in the loader.

If that is so, then is the `legacy_save` parameter in `save` unnecessary? Or will it load only files saved in the new way now? If that does not "match", then this comment is blocking (as it would mean users cannot use `legacy_save` for loading, saving with that param becomes "write-only").

**Comment by benHeid** [comment]
> Approved under the assumption that you tested it.
> 
> 
> 
> What I'm a bit surprised about is that there's no `legacy_save` method distinction in the loader.
> 
> 
> 
> If that is so, then is the `legacy_save` parameter in `save` unnecessary? Or will it load only files saved in the new way now? If that does not "match", then this comment is blocking (as it would mean users cannot use `legacy_save` for loading, saving with that param becomes "write-only").

Let me test it again if the old save and load again works. But I think I run the test with the new and an older tensorflow version. If both work then everything should be fine or?

**Comment by fkiraly** [comment]
> If both work then everything should be fine or?

Actually, what I was thinking, new `tensorflow` version with `legacy_save` on and off, and then try `load` twice. Btw, are we testing this somewhere?

**Comment by benHeid** [comment]
This would result in the same execution path in the CI, since the legacy safe is only considered if the version is lower than 2.16.0. However in the CI pipeline we are installing the latest version

**Comment by fkiraly** [comment]
So, then my question is, why do we need `legacy_save` at all - what user code is impacted? If `load` work the same whether this flag is set or not, then there's no user code we're breaking, and no change/deprecation is needed.

Or am I getting sth wrong?

File: sktime/classification/tests/test_base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yarnabrina) [reply sktime/classification/tests/test_base.py:503]: @benHeid sorry to opening this resolved question once more, but I see tests do not have to use the latest verions always if triggered from `classification` or `network` or `regression` (or `dl` - won't happen in new CI). Those have upper bounds set as 2.14, probably based on latest version at time of creation (though why inclusive I can't remember).

If those upper bounds are no longer necessary, may be those can be updated to recent ones.

**Comment by fkiraly** [comment]
Ah, I see. Yes, that makes sense.

**Comment by benHeid** [comment]
> * will this then imply a new lower bound on tensorflow? I would guess not, since the new behaviour makes a version distinction

No, we do not need a new lower bound. Old versions and the new version are working.

> * "it might be the case that it breaks if the model is stored differently" - is this really the case, if `load` works no matter what?
I haven't tested it. My assumption is: From sktime side, it should work since I just look which files are available and depending on that the model is loaded. However, I am unsure if the new tensorflow version would load models that are stored using and older version.

File: sktime/classification/tests/test_base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (benHeid) [reply sktime/classification/tests/test_base.py:503]: Oh okay... This explains why this failure is only occurring on mlflow and pandas 1. In these cases, tensorflow 2.16.1 is installed.  I can try to execute everything without that bound. -> All extras is installing tensorflow without any bounds. This seems to be a bit inconsistent.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Hm, with the version bound removed, it fails now.

I'll make a separate PR that adds bounds `tf<2.6.0`

**Comment by fkiraly** [comment]
Here: https://github.com/sktime/sktime/pull/6117

**Comment by benHeid** [comment]
It seems that the errors are related to the save/load of the deep learning based regressors.. I try to fix them

**Review by fkiraly — APPROVED** [review]
I see - looks like they are fixed now!

To confirm quickly, this means we no longer need #6117 , right?

**Comment by yarnabrina** [comment]
> I see - looks like they are fixed now!
> 
> To confirm quickly, this means we no longer need #6117 , right?

I'm very sorry, I didn't see this and merged that one to not see failures in other PR's. Now that leads to conflicts in this one, apologies.

**Comment by benHeid** [comment]
Should I resolve the conflicts or do should we revert the other PR?

**Comment by fkiraly** [comment]
> Should I resolve the conflicts or do should we revert the other PR?

I would say, the clean way is resolving the conflicts *by* reverting the other PR as part of this one.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Makes sense.

Minor things, non-blocking (I can fix these):

* warning and effective path check should be located in the public function
* typo in warning, legacy_save, not legacy_warning
* duplication depending on serialization_format, this is not introduced in this PR but impacted. I would refactor it to a single call of `_serialize_using_dump_func`, the branch affecting only the package that is being referenced. I.e., `if serialization_format == "cloudpickle": serial_pkg = cloudpickle`

**Review by fkiraly — APPROVED** [review]

**Pull request merged**